From c4c6fadd3d61c6d66ba6026791571fb394ac27e2 Mon Sep 17 00:00:00 2001
From: Eric Naim <dnaim@cachyos.org>
Date: Thu, 18 Dec 2025 12:36:06 +0800
Subject: [PATCH 3/3] Fix compile for 6.19

Contains:
- Rename page_free callback -> folio_free callback for 6.19+
- Adjust zone_device_page_init() call for 6.19; it has one extra argument now

Signed-off-by: Eric Naim <dnaim@cachyos.org>
---
 kernel-open/nvidia-uvm/uvm_hmm.c     |  4 ++++
 kernel-open/nvidia-uvm/uvm_pmm_gpu.c | 34 ++++++++++++++++++++++++++++
 2 files changed, 38 insertions(+)

diff --git a/kernel-open/nvidia-uvm/uvm_hmm.c b/kernel-open/nvidia-uvm/uvm_hmm.c
index 3e995e2dec76..5daf91ecb9b8 100644
--- a/kernel-open/nvidia-uvm/uvm_hmm.c
+++ b/kernel-open/nvidia-uvm/uvm_hmm.c
@@ -2146,7 +2146,11 @@ static void fill_dst_pfn(uvm_va_block_t *va_block,
 
         UVM_ASSERT(!page_count(dpage));
         UVM_ASSERT(!dpage->zone_device_data);
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 19, 0)
+        zone_device_page_init(dpage, 0);
+#else
         zone_device_page_init(dpage);
+#endif
         dpage->zone_device_data = gpu_chunk;
         atomic64_inc(&va_block->hmm.va_space->hmm.allocated_page_count);
     }
diff --git a/kernel-open/nvidia-uvm/uvm_pmm_gpu.c b/kernel-open/nvidia-uvm/uvm_pmm_gpu.c
index cba0ec5397fe..191c26b47790 100644
--- a/kernel-open/nvidia-uvm/uvm_pmm_gpu.c
+++ b/kernel-open/nvidia-uvm/uvm_pmm_gpu.c
@@ -175,6 +175,8 @@
 #include "uvm_test.h"
 #include "uvm_linux.h"
 
+#include <linux/version.h>
+
 #if defined(CONFIG_PCI_P2PDMA) && defined(NV_STRUCT_PAGE_HAS_ZONE_DEVICE_DATA)
 #include <linux/pci-p2pdma.h>
 #endif
@@ -3115,8 +3117,14 @@ static bool uvm_pmm_gpu_check_orphan_pages(uvm_pmm_gpu_t *pmm)
     return ret;
 }
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 19, 0)
+static void devmem_folio_free(struct folio *folio)
+{
+    struct page *page = &folio->page;
+#else
 static void devmem_page_free(struct page *page)
 {
+#endif
     uvm_gpu_chunk_t *chunk = uvm_pmm_devmem_page_to_chunk(page);
     uvm_gpu_t *gpu = uvm_gpu_chunk_get_gpu(chunk);
 
@@ -3177,7 +3185,11 @@ static vm_fault_t devmem_fault_entry(struct vm_fault *vmf)
 
 static const struct dev_pagemap_ops uvm_pmm_devmem_ops =
 {
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 19, 0)
+    .folio_free = devmem_folio_free,
+#else
     .page_free = devmem_page_free,
+#endif
     .migrate_to_ram = devmem_fault_entry,
 };
 
@@ -3265,8 +3277,14 @@ static void device_p2p_page_free_wake(struct nv_kref *ref)
     wake_up(&p2p_mem->waitq);
 }
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 19, 0)
+static void device_p2p_folio_free(struct folio *folio)
+{
+    struct page *page = &folio->page;
+#else
 static void device_p2p_page_free(struct page *page)
 {
+#endif
     uvm_device_p2p_mem_t *p2p_mem = page->zone_device_data;
 
     page->zone_device_data = NULL;
@@ -3275,14 +3293,26 @@ static void device_p2p_page_free(struct page *page)
 #endif
 
 #if UVM_CDMM_PAGES_SUPPORTED()
+
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 19, 0)
+static void device_coherent_folio_free(struct folio *folio)
+{
+    device_p2p_folio_free(folio);
+}
+#else
 static void device_coherent_page_free(struct page *page)
 {
     device_p2p_page_free(page);
 }
+#endif
 
 static const struct dev_pagemap_ops uvm_device_coherent_pgmap_ops =
 {
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 19, 0)
+    .folio_free = device_coherent_folio_free,
+#else
     .page_free = device_coherent_page_free,
+#endif
 };
 
 static NV_STATUS uvm_pmm_cdmm_init(uvm_parent_gpu_t *parent_gpu)
@@ -3419,7 +3449,11 @@ static bool uvm_pmm_gpu_check_orphan_pages(uvm_pmm_gpu_t *pmm)
 
 static const struct dev_pagemap_ops uvm_device_p2p_pgmap_ops =
 {
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 19, 0)
+    .folio_free = device_p2p_folio_free,
+#else
     .page_free = device_p2p_page_free,
+#endif
 };
 
 void uvm_pmm_gpu_device_p2p_init(uvm_parent_gpu_t *parent_gpu)
-- 
2.52.0

